### PR TITLE
C++ API Bindings for Heron.

### PR DESCRIPTION
**Pull request by srkukarni**
Based on design doc at
https://docs.google.com/document/d/1fpLuwEXtLWAxhL9MCI_EsxQ3bYBcdXB7G8xMguF35LM/edit#heading=h.d6uy2uxfs2xq

### CODE REVIEW

**Comment by objmagic** [comment]
Two general comments:
1. There is no unit test/integration test. Is this PR done?
2. I think when this PR is ready to be merged, I will run an internal integration test on it.

**Comment by srkukarni** [comment]
@objmagic This is just the C++ api. Mostly interfaces. Integration tests will come after the cpp-instance is merged and examples are created.

**Comment by objmagic** [comment]
@srkukarni Understood. Ping me when you are ready for an internal integration test.

**Comment by huijunw** [comment]
related #1512, #1513, #1514, #1515

File: heron/api/src/cpp/bolt/base-basic-bolt.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#include "config/config.h"
+#include "bolt/ibasic-bolt.h"
+
+namespace heron {
+namespace api {
+namespace bolt {
+
+class BaseBasicBolt : public IBasicBolt {
+ public:
+  virtual void prepare(std::shared_ptr<config::Config> conf,
```
Comment (srkukarni) [inline heron/api/src/cpp/bolt/base-basic-bolt.h]: Added

File: heron/api/src/cpp/bolt/base-basic-bolt.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                       std::shared_ptr<topology::TaskContext> context) {
+  }
+
+  virtual void cleanup() {
+  }
+
+  virtual std::shared_ptr<config::Config> getComponentConfiguration() {
+    return NULL;
```
Comment (srkukarni) [inline heron/api/src/cpp/bolt/base-basic-bolt.h]: Changed everywhere

File: heron/api/src/cpp/bolt/irich-bolt.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "bolt/ibolt.h"
+#include "topology/icomponent.h"
+
+namespace heron {
+namespace api {
+namespace bolt {
+
+/**
+ * IRichBolt and IRichBolt are the main interfaces to use to implement components of the topology.
```
Comment (srkukarni) [inline heron/api/src/cpp/bolt/irich-bolt.h]: Corrected

**Pull request closed without merging**